backend/metadata_writer/src/metadata_writer.py

```diff
@@ -0,0 +1,333 @@
+
+
+def output_name_to_argo(name: str) -> str:
+    import re
+    return re.sub('-+', '-', re.sub('[^-0-9a-z]+', '-', name.lower())).strip('-')
+
+
+def argo_artifact_to_uri(artifact: dict) -> str:
+    if 's3' in artifact:
+        s3_artifact = artifact['s3']
+        return 'minio://{bucket}/{key}'.format(
+            bucket=s3_artifact.get('bucket', ''),
+            key=s3_artifact.get('key', ''),
+        )
+    elif 'raw' in artifact:
+        return None
+    else:
+        return None
+
+
+def is_tfx_pod(pod) -> bool:
+    main_containers = [container for container in pod.spec.containers if container.name == 'main']
+    if len(main_containers) != 1:
+        return False
+    main_container = main_containers[0]
+    return main_container.command and main_container.command[-1].endswith('tfx/orchestration/kubeflow/container_entrypoint.py')
+
+
+# Caches (not expected to be persistent)
+# These caches are only used to prevent race conditions. Race conditions happen because the writer can see multiple versions of K8s object before the applied labels show up.
+# They are expected to be lost when restarting the service.
+# The operation of the Metadata Writer remains correct even if it's getting restarted frequently. (Kubernetes only sends the latest version of resource for new watchers.)
+# Technically, we could remove the objects from cache as soon as we see that our labels have been applied successfully.
+pod_name_to_execution_id = {}
+workflow_name_to_context_id = {}
+pods_with_written_metadata = set()
+
+while True:
+    print("Start watching Kubernetes Pods created by Argo")
+    for event in k8s_watch.stream(
+        k8s_api.list_namespaced_pod,
+        namespace=namespace_to_watch,
+        label_selector=ARGO_WORKFLOW_LABEL_KEY,
+    ):
+        try:
+            obj = event['object']
+            print('Kubernetes Pod event: ', event['type'], obj.metadata.name, obj.metadata.resource_version)
+            if event['type'] == 'ERROR':
+                print(event)
+
+            # Logging pod changes for debugging
+            try:
+                import yaml
+                with open('/tmp/pod_' + obj.metadata.name + '_' + obj.metadata.resource_version, 'w') as f:
+                    f.write(yaml.dump(obj.to_dict()))
+            except:
+                pass
+            assert obj.kind == 'Pod'
+
+            if METADATA_WRITTEN_LABEL_KEY in obj.metadata.labels:
+                continue
+
+            # Skip TFX pods - they have their own metadata writers
+            if is_tfx_pod(obj):
+                continue
+
+            argo_workflow_name = obj.metadata.labels[ARGO_WORKFLOW_LABEL_KEY] # Should exist due to initial filtering
+            argo_template = json.loads(obj.metadata.annotations[ARGO_TEMPLATE_ANNOTATION_KEY])
+            argo_template_name = argo_template['name']
+
+            component_name = argo_template_name
+            component_version = component_name
```

Comment:
Is version = name?

Reply:
Never mind, I see you updated it below.

Reply:
I should have probably called it `versioned_name` to avoid confusion.